openedx_tagging/core/tagging/api.py

```diff
@@ -107,3 +107,73 @@ def tag_object(
     """
 
     return taxonomy.tag_object(tags, object_id, object_type)
+
+
+def autocomplete_tags(taxonomy: Taxonomy, prefix: str, object_id: str= None, count=10) -> List[TagResult]:
+    """
+    Returns the first `count` tag values in the given Taxonomy with names 
+    that begin with the given prefix string. The result is returned in alphabetical
+
+    Closed taxonomies return tag values that exist in the taxonomy. Also excludes all Tags that the
+    `object_id` already has. 
+
+    Free-text taxonomies return only tag values that are currently in use on (another) object. Also excludes
+    the tags used by `object_id`.
+    """
+    result = []
+
+    # Fetch tags that the object already has to exclude them from the result
+    excluded_tags = ObjectTag.objects.filter(object_id=object_id, taxonomy=taxonomy)
+    if taxonomy.allow_free_text:
+        # Free-text taxonomy
+
+        # Obtain the value of the excluded tags
+        excluded_tags = excluded_tags.values_list('_value', flat=True)
+        result = (
+            # Fetch object tags from this taxonomy whose value starts with the given prefix
+            ObjectTag.objects.filter(taxonomy=taxonomy, _value__istartswith=prefix)
+            # omit any tags applied to the given object
+            .exclude(object_id=object_id)
+            # omit any free-text tags from other objects whose values match the tags on the given object
+            .exclude(_value__in=excluded_tags)
+            # alphabetical ordering
+            .order_by('_value')
+            # obtain the values of the tags
+            .values_list('_value', flat=True)
+            # remove repeats
```

Comment:
In what scenarios could we get repeats?

Reply:
About the free-form taxonomies, we do the search in ObjectTag. There may be the scenario where different objects have the same tag